core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/JoinClause.kt

```diff
@@ -21,43 +21,75 @@ private enum class JoinType(val type: String) {
-        val parentDopeQuery = parentClause.toDopeQuery()
-        val bucketDopeQuery = bucket.toDopeQuery()
-        val keyDopeQuery = onKey.toDopeQuery()
-        val forBucketDopeQuery = forBucket.toDopeQuery()
-        dopeQuery = DopeQuery(
-            queryString = "${parentDopeQuery.queryString} ${joinType.type} ${bucketDopeQuery.queryString} " +
-                "ON KEY ${keyDopeQuery.queryString} FOR ${forBucketDopeQuery.queryString}",
-            parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + keyDopeQuery.parameters + forBucketDopeQuery.parameters,
-        )
+        this.joinType = joinType
+        this.bucket = bucket
+        this.onKey = onKey
+        this.forBucket = forBucket
+        this.parentClause = parentClause
     }
 
-    override fun toDopeQuery() = dopeQuery
+    override fun toDopeQuery(): DopeQuery =
+        when {
+            onCondition != null -> {
+                val parentDopeQuery = parentClause.toDopeQuery()
+                val bucketDopeQuery = bucket.toDopeQuery()
+                val onConditionDopeQuery = onCondition!!.toDopeQuery()
+                DopeQuery(
+                    queryString = "${parentDopeQuery.queryString} ${joinType.type} ${bucketDopeQuery.queryString} " +
+                        "ON ${onConditionDopeQuery.queryString}",
+                    parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + onConditionDopeQuery.parameters,
+                )
+            }
+
+            onKeys != null -> {
+                val parentDopeQuery = parentClause.toDopeQuery()
+                val bucketDopeQuery = bucket.toDopeQuery()
+                val keyDopeQuery = onKeys!!.toDopeQuery()
+                DopeQuery(
+                    queryString = "${parentDopeQuery.queryString} ${joinType.type} ${bucketDopeQuery.queryString} " +
+                        "ON KEYS ${keyDopeQuery.queryString}",
+                    parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + keyDopeQuery.parameters,
+                )
+            }
+
+            onKey != null && forBucket != null -> {
+                val parentDopeQuery = parentClause.toDopeQuery()
+                val bucketDopeQuery = bucket.toDopeQuery()
+                val keyDopeQuery = onKey!!.toDopeQuery()
+                val forBucketDopeQuery = forBucket!!.toDopeQuery()
+                DopeQuery(
+                    queryString = "${parentDopeQuery.queryString} ${joinType.type} ${bucketDopeQuery.queryString} " +
+                        "ON KEY ${keyDopeQuery.queryString} FOR ${forBucketDopeQuery.queryString}",
+                    parameters = parentDopeQuery.parameters + bucketDopeQuery.parameters + keyDopeQuery.parameters +
+                        forBucketDopeQuery.parameters,
+                )
+            }
+
+            else -> throw IllegalStateException("Unable to construct join clause")
```

Comment:
When can this be the case? Please be more specific in the error-message

Reply:
This is thrown, when some somehow created a Join Clause with a null value, that shouldn't and can't be null.
What I mean is, that this realy only happens, when you do something impossible

Reply:
Ah I see you made a "hacky" version to differenciate between the three constructors. This is not a very good way, first you make a lot assumptions (e.g. if there is a _onCondition_ ther is no _onkey_)
Can't you build it dynamically together depending on which field is set?

Reply:
I don't have an idea how to implement it in an other way except nullable parameters in one constructor, which wouldn't solve the problem. Do you have an idea?